modules/self_snapshot/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+#![no_std]
+#![feature(
+    core_intrinsics,
+    lang_items,
+    alloc_error_handler,
+    arbitrary_self_types
+)]
+
+use dallo::{HostAlloc, ModuleId, State};
+#[global_allocator]
+static ALLOCATOR: HostAlloc = HostAlloc;
+
+const ARGBUF_LEN: usize = 64;
+
+#[no_mangle]
+static mut A: [u64; ARGBUF_LEN / 8] = [0; ARGBUF_LEN / 8];
+#[no_mangle]
+static AL: i32 = ARGBUF_LEN as i32;
+
+#[no_mangle]
+static SELF_ID: ModuleId = ModuleId::uninitialized();
+
+static mut STATE: State<SelfSnapshot> =
+    unsafe { State::new(SelfSnapshot { crossover: 7 }, &mut A) };
+
+pub struct SelfSnapshot {
+    crossover: i32,
+}
+
+impl SelfSnapshot {
+    pub fn crossover(&self) -> i32 {
+        self.crossover
+    }
+
+    pub fn set_crossover(&mut self, to: i32) -> i32 {
+        let old_val = self.crossover;
+        // dallo::debug!(
+        //     "setting crossover from {:?} to {:?}",
+        //     self.crossover,
+        //     to
+        // );
+        self.crossover = to;
+        old_val
+    }
+
+    // updates crossover and returns the old value
+    pub fn self_call_test_a(self: &mut State<Self>, update: i32) -> i32 {
```

Comment:
I'm confused by the `self` here, why does it have a type, if so, that means, it is not a real self? Maybe naming it differently would be less confusing?

Reply:
The self type is for tying the lifetime of the self reference to the wrapper struct `State`, to leverage the borrow checker for self-calling. 

In essence it guarantees that when  transacting on `dallo::self_id()` no other aliases or borrows of `SELF` exist.

Reply:
This keeps the call structure consistent, since a `&mut self` is freely interchangable if no calling is needed for the execution.